src/profile-logic/line-timings.js

```diff
@@ -217,6 +217,219 @@ export function getStackLineInfoInverted(
+ *     For stacks that map to the given call node or one of its descendent
+ *     call nodes, this is a set containing one element, which is
+ *     ancestorStack.frame.line, where ancestorStack maps to the given call
+ *     node.
+ *     For all other stacks, this is null.
+ */
+export function getStackLineInfoForCallNodeNonInverted(
+  stackTable: StackTable,
+  frameTable: FrameTable,
+  callNodeIndex: IndexIntoCallNodeTable,
+  { stackIndexToCallNodeIndex }: CallNodeInfo
+): StackLineInfo {
+  // "self line" == "the line which a stack's self time is contributed to"
+  const callNodeSelfLineForAllStacks = [];
+  // "total lines" == "the set of lines whose total time this stack contributes to"
+  const callNodeTotalLinesForAllStacks = [];
+
+  // This loop takes advantage of the fact that the stack table is topologically ordered:
+  // Prefix stacks are always visited before their descendants.
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    let selfLine: LineNumber | null = null;
+    let totalLines: Set<LineNumber> | null = null;
+
+    if (stackIndexToCallNodeIndex[stackIndex] === callNodeIndex) {
+      // This stack contributes to the call node's self time.
+      // We don't need to check the stack's func or file because it'll be
+      // the same as the given call node's func and file.
+      const frame = stackTable.frame[stackIndex];
+      selfLine = frameTable.line[frame];
+      if (selfLine !== null) {
+        totalLines = new Set([selfLine]);
+      }
+    } else {
+      // Inherit total time contribution from the parent stack. selfLine is null.
+      const prefixStack = stackTable.prefix[stackIndex];
+      totalLines =
+        prefixStack !== null
+          ? callNodeTotalLinesForAllStacks[prefixStack]
+          : null;
+    }
+
+    callNodeSelfLineForAllStacks.push(selfLine);
+    callNodeTotalLinesForAllStacks.push(totalLines);
+  }
+  return {
+    selfLine: callNodeSelfLineForAllStacks,
+    stackLines: callNodeTotalLinesForAllStacks,
+  };
+}
+
+/**
+ * This handles the inverted case of getStackLineInfoForCallNode.
+ *
+ * The returned StackLineInfo is computed as follows:
+ *   selfLine[stack]:
+ *     For (inverted thread) root stack nodes that map to the given call node
+ *     and whose stack.frame.func.file is the given file, this is stack.frame.line.
+ *     For (inverted thread) root stack nodes whose frame is in a different file,
+ *     or which don't map to the given call node, this is null.
+ *     For (inverted thread) *non-root* stack nodes, this is the same as the selfLine
+ *     of the stack's prefix. This way, the selfLine is always inherited from the
+ *     subtree root.
+ *   stackLines[stack]:
+ *     For stacks whose stack.frame.func.file is the given file, this is the stackLines
+ *     of its (inverted thread) prefix stack, plus stack.frame.line added to the set.
+ *     For all other stacks this is the same as the stackLines set of the stack's prefix.
```

Comment:
While tweaking the comments I noticed that this comment was completely wrong. Will fix.